### ISSUE TITLE
SDPA bad PCC for Qwen 2.5 shapes and inputs

### ISSUE THREAD

**Issue by yieldthought** [tenstorrent/tt-metal#21199]
### Describe the bug

Qwen 2.5 is showing very degraded output at long sequence lengths. We already see the SDPA prefill op giving bad PCC if I use the actual Qwen input tensor data:

```
Q: -69.0 < 0.1 ± 3.5 < 66.0 - torch.Size([1, 28, 4096, 128])
K: -166.0 < 1.3 ± 26.8 < 175.0 - torch.Size([1, 4, 4096, 128])
V: -3.0 < 0.0 ± 0.3 < 3.0 - torch.Size([1, 4, 4096, 128])

python vs pytorch: Max ATOL Delta: 2.984375, Max RTOL Delta: inf, PCC: 0.9179217348935194, PCC check failed
mse: TorchTensor(0.0028)
```

If I use the same shapes with the pytest's default random input (mean 0, std 1) the test passes. If I scale the inputs to match the above mean and std it fails, but with 0.98 PCC instead of 0.91 PCC. The real values seem to matter a lot to the accuracy.

**Comment by cglagovichTT** [comment]
@yieldthought do you have a unit test I can use to debug this

**Comment by yieldthought** [comment]
I'll send you the QKV files directly, too large to upload here.

**Comment by yieldthought** [comment]
https://github.com/tenstorrent/tt-inference-server/issues/125

**Comment by cglagovichTT** [comment]
Tracking Mark's failing test and debug progress on branch https://github.com/tenstorrent/tt-metal/tree/cglagovich/21199

Reproed poor PCC with test
```bash
pytest tests/tt_eager/python_api_testing/unit_testing/misc/test_scaled_dot_product_attention.py::test_sdpa_tt -k '4096 and bfp8 and q128 and k512' --tb=short
```
`python vs pytorch: Max ATOL Delta: 2.8125, Max RTOL Delta: inf, PCC: 0.9470340284025829, PCC check failed`

Attempted improving `reduce_sum` fidelity by swapping with a matmul implementation:
`python vs pytorch: Max ATOL Delta: 2.8125, Max RTOL Delta: inf, PCC: 0.9470784154390033, PCC check failed`

Next steps
- Change statistics CBs to FP32 to preserve precision

**Comment by yieldthought** [comment]
Qwen 7B's QKV bias values are crazy high, which is producing these very large activation ranges going into SDPA:

```
>>> print([f'{k:.0f}' for k in K[0, 0, 0].tolist()])
['-1', '-4', '2', '-0', '1', '1', '-0', '-2', '0', '-1', '0', '-1', '2', '1', '1', '-0', '0', '-0', '-1', '0', '0', '-1', '0', '0', '3', '1', '-1', '-1', '-1', '-0', '-0', '0', '-1', '-1', '0', '0', '1', '-0', '0', '3', '0', '0', '0', '0', '4', '-2', '-0', '0', '0', '0', '1', '-0', '0', '-0', '-0', '0', '-0', '1', '0', '0', '1', '-1', '4', '4', '-0', '0', '-0', '1', '-0', '-1', '0', '-1', '-0', '1', '1', '-1', '-1', '-0', '1', '-0', '1', '-1', '-8', '13', '-1', '-1', '-1', '-1', '-3', '3', '-1', '-0', '-0', '-1', '-1', '-5', '-1', '-3', '11', '10', '-0', '-2', '-16', '-2', '-14', '16', '-22', '-10', '20', '-46', '-71', '16', '76', '84', '104', '70', '78', '-2', '70', '162', '78', '158', '132', '-116', '-106', '-164', '172', '-114']
```

This comes directly from the bias:
```python
>>> for model in [3, 7, 72]: print(f'{model}B:', AutoModelForCausalLM.from_pretrained(f'Qwen/Qwen2.5-{model}B-Instruct').model.layers[0].self_attn.k_proj.bias.std().item())
3B: 15.369823455810547
7B: 26.71364402770996
72B: 3.170764684677124
```

Qwen 3 does not appear to use bias terms in QKV or the MLP.

**TL;DR:** The bad Qwen 2.5 7B output appears to be caused by numerical accuracy issues triggered by extremely large bias values in the K projection. Other Qwen 2.5 models, in particular the flagship 72B model, do not have such large bias values and Qwen 3 in common with most other architectures does not use QKV bias terms at all.

It's worth considering whether we should focus our efforts on Qwen 3 bringup and mark Qwen 2.5 7B down as a known issue due to its large bias values.

**Comment by cglagovichTT** [comment]
Making very strong progress. 

Tested:
- fp32 MM intermediates
- fp32 statistics CBs
- exp-approx for stats and softmax
- fp32 scalar multiple
- stochastic rounding on/off with fp32_acc on/off
- fp32 inputs/outputs

The only thing that moved the needle was fp32 inputs/outputs. I achieved `PCC 0.9998056813572749, MSE 7.169172022258863e-06` with fp32 in/out. All other tests yielded `PCC < 0.9685435653406544`.

Next steps:
- ablate all changes to find minimal set required for `PCC >= 0.999` on Qwen inputs

**Comment by cglagovichTT** [comment]
I found that PCC is too low (<.968) when Q input is bfloat16. 
When Q input is fp32 and K, V, output are bfloat16, PCC is acceptable at 0.9998.

In theory there is no reason why Q input must be fp32. The reference input tensor is bf16, so upcasting to fp32 does not increase precision. The only hint we have is that an L1 dtype of fp32 configures the unpacker register to tf32, though we do not understand how this could affect correctness because there are still only 16 bits of information in the tf32 tile register. 

We have 2 hypotheses why Q must be fp32 to get good outputs. 
1. There is some issue with MM configuration that is worked around with Q in fp32
2. Against our understanding, performing MM with a tf32 srcB register modifies FPU execution in some way that impacts correctness.

Next steps:
- @cglagovichTT pushes branch with minimal changes to achieve >.9998 PCC on Qwen tensors
- @cglagovichTT opens P2 issue to understand the root cause of this problem
- @yieldthought tests Qwen with SDPA branch

**Comment by uaydonat** [comment]
check with @rtawfik01 - if srcB is tf32 does that automatically change srcA type as well?

**Comment by cglagovichTT** [comment]
@yieldthought please try Qwen tests with this branch https://github.com/tenstorrent/tt-metal/compare/main...cglagovich/21199_sdpa_correctness

Use this compute kernel config
```python
    compute_kernel_config = ttnn.WormholeComputeKernelConfig(
        math_fidelity=ttnn.MathFidelity.HiFi4,
        math_approx_mode=False,
        fp32_dest_acc_en=True,
        packer_l1_acc=False,
    )
```

Ensure you use fp32 matmul intermediates in your program config
```python
    program_config = ttnn.SDPAProgramConfig(
        compute_with_storage_grid_size=device.compute_with_storage_grid_size(),
        q_chunk_size=q_chunk_size,
        k_chunk_size=k_chunk_size,
        exp_approx_mode=False,
        fp32_matmul_intermediates=True,
    )
```

and set Q dtype to float32.

I reduced the `q_chunk_size` to 128 and kept `k_chunk_size` at 256 to avoid OOM L1.

**Comment by uaydonat** [comment]
closing since we have a solution
